### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by jschofield1**
Collaborated with @coproduct

### CODE REVIEW

**Comment by jschofield1** [comment]
1.	Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
•	Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
•	Using LINQ, skip the first row in the People.csv. ✔
•	Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
1.	Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
•	Use ISampleData.CsvRows for your data source. ✔
•	Don't forget the list should be unique. ✔
•	Sort the list alphabetically. ✔
•	Include a test that uses LINQ to verify the data is sorted correctly (do not use a hard coded list). ✔
1.	Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
•	Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
•	Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
•	Given the array, consider using string.Join to combine the list into a single string. ✔
1.	Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
•	Use ISampleData.CsvRows as the source of the data. ✔
•	Sort the list by State, City, Zip. (Sort the addresses first then select). ✔
•	Be sure that Person.Address is also populated. ✔
•	Adding null validation to all the Person and Address properties is optional.
•	Consider using ISampleData.CsvRows in your test to verify your results. ✔
1.	Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
•	Use ISampleData.People for your data source. ✔
1.	Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma separated list of states. ✔
•	Use the people parameter from ISampleData.GetUniqueListOfStates for your data source. ✔
•	At a minimum, use System.Linq.Enumerable.Aggregate LINQ method to create your result. ✔
•	Don't forget the list should be unique. ✔
•	It is recommended that at a minimum you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
1.	Given the implementation of Node in Assignment5
•	Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
•	Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
Extra Credit
The following are options for extra credit (you don't need to do them all):
•	Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements ISampleData). Refactor both your SampleData and SampleDataAsync classes so there is minimal duplication. This includes refactoring your tests so a significant amount of test code can be re-used. ❌
Fundamentals
•	Ensure you enable nullable reference types, net5 targeted, C# 9.0, warnings as errors, and enabled .NET analyzers for both projects ✔
•	Fully unit test the assignment ✔
•	Choose simplicity over than complexity ✔

Everything seems to be working properly and following proper naming conventions.

File: Assignment/Assignment/SampleData.cs [hunk 1]
```diff
@@ -1,30 +1,79 @@
...
+            {
+                List<string> result = new List<string>();
+                using (var fileReader = new StreamReader(@$".\People.csv"))
+                {
+                    for (string? s = fileReader.ReadLine(); !fileReader.EndOfStream; s = fileReader.ReadLine())
+                    {
+                        result.Add(s!);
+                    }
+                }
+                return result.Skip<string>(1);
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:31]: Consider using `File.ReadLines` or `File.ReadAllLines` here. There is no need to use a for loop.

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,79 @@
...
+            }
+        }
 
         // 2.
         public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() 
-            => throw new NotImplementedException();
+        {
+            SampleData sampleData = new SampleData();
+            IEnumerable<IPerson> personCollection = sampleData.People;
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:39]:
```suggestion
            IEnumerable<IPerson> personCollection = People;
```
You don't need to create a new instance of the class if you're in that class.

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,79 @@
...
+                    {
+                        result.Add(s!);
+                    }
+                }
+                return result.Skip<string>(1);
+            }
+        }
 
         // 2.
         public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:36]: This method should be using `CsvRows` as its data source, not `People`

File: Assignment/Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,79 @@
...
+        }
 
         // 2.
         public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() 
-            => throw new NotImplementedException();
+        {
+            SampleData sampleData = new SampleData();
+            IEnumerable<IPerson> personCollection = sampleData.People;
+
+            IEnumerable<string> statesList = personCollection.Select(item => item.Address.State).Distinct<string>().OrderBy(s => s);
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:41]:
```suggestion
            return personCollection.Select(item => item.Address.State).Distinct().OrderBy(s => s);
```

File: Assignment/Assignment/SampleData.cs [hunk 5]
```diff
@@ -1,30 +1,79 @@
...
+
+            return statesList;
+        }
 
         // 3.
         public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+        {
+            SampleData sample = new SampleData();
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:49]:
```suggestion
```

File: Assignment/Assignment/SampleData.cs [hunk 6]
```diff
@@ -1,30 +1,79 @@
...
+            IEnumerable<string> stateCollection = sample.GetUniqueSortedListOfStatesGivenCsvRows();
+            string[] stateStringArray = stateCollection.ToArray<string>();
+            return String.Join(", ", stateStringArray);
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:52]: This can be condensed to one line.

Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:52]:
```suggestion
            IEnumerable<string> stateCollection = sample.GetUniqueSortedListOfStatesGivenCsvRows();
            string[] stateStringArray = stateCollection.ToArray<string>();
            return string.Join(", ", stateStringArray);
```
Do not use BCL names.

File: Assignment/Assignment/SampleData.cs [hunk 7]
```diff
@@ -1,30 +1,79 @@
...
+        }
 
         // 4.
-        public IEnumerable<IPerson> People => throw new NotImplementedException();
+        public IEnumerable<IPerson> People => CsvRows.Select(item =>
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:56]: You should be ordering first.

File: Assignment/Assignment/SampleData.cs [hunk 8]
```diff
@@ -1,30 +1,79 @@
...
 {
     public class SampleData : ISampleData
     {
+        const int FirstName = 1;
+        const int LastName = 2;
+        const int EmailAddress = 3;
+        const int StreetAddress = 4;
+        const int City = 5;
+        const int State = 6;
+        const int Zip = 7;
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:16]: Why not use an enum?

File: Assignment/Assignment/SampleData.cs [hunk 9]
```diff
@@ -1,30 +1,79 @@
...
+        {
+            return from Person in People where filter(Person.EmailAddress) select (Person.FirstName, Person.LastName);
+        }
 
         // 6.
         public string GetAggregateListOfStatesGivenPeopleCollection(
-            IEnumerable<IPerson> people) => throw new NotImplementedException();
+            IEnumerable<IPerson> people)
+        {
+            return people.Select(item => item.Address.State).Distinct().Aggregate((result, next) => result + ", " + next);
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:76]: Consider using string interpolation.

**Review by mmwoodfo — COMMENTED** [review]
I can't find a code review by you. Did you do one?

**Comment by jschofield1** [comment]
> 
> 
> @jschofield1 I can't find a code review by you. Did you do one?

Yes, you should find a self review in comment number 2 of the pull request. Peer review was optional since @coproduct and I partnered up.

**Comment by mmwoodfo** [comment]
@jschofield1 Got it, sorry. I think this is the second time now that I missed noticing you paired on the assignment. :)

**Comment by jschofield1** [comment]
No worries!
________________________________

**Pull request closed without merging**